### PR TITLE
Quickstart

### PR DESCRIPTION
**Pull request by foreign-sub**
* fix quickstart.md
* test quickstart in ci

### CODE REVIEW

File: docs/quickstart.md [hunk 1]
```diff
@@ -101,17 +119,33 @@ class TinyNet:
...
+    x = self.l1(x).leakyrelu()
+    return self.l2(x).log_softmax()
 
 net = TinyNet()
+
 ```
+
 We can see that the forward pass of our neural network is just the sequence of operations performed on the input tensor `x`.
-We can also see that functional operations like `leakyrelu` and `log_softmax` are not defined as classes and instead are just methods we can just call.
+We can also see that functional operations like `leakyrelu` and `log_softmax` are not defined as classes and instead are just methods we can just call like that or with `x = x.leakyrelu()`.
```
Comment (wozeparrot) [inline docs/quickstart.md:130]: This line becomes a little confusing with this change.

File: docs/quickstart.md [hunk 2]
```diff
@@ -135,46 +173,76 @@ def cross_entropy(out, Y):
...
 Namely, operations that are load/store like indexing a tensor with another tensor or assigning a value to a tensor at a certain index.
 Load/store ops are not supported in tinygrad because they add complexity when trying to port to different backends and 90% of the models out there don't use/need them.
 
 For our optimizer we will be using the traditional stochastic gradient descent optimizer with a learning rate of 3e-4.
+
 ```py
 from tinygrad.nn.optim import SGD
 
-opt = SGD([net.l1.weight, net.l2.weight], lr=3e-4)
+#opt = SGD([net.l1.weight, net.l2.weight], lr=3e-4)
```
Comment (wozeparrot) [inline docs/quickstart.md:188]: Why was this commented out?

File: docs/quickstart.md [hunk 3]
```diff
@@ -135,46 +173,76 @@ def cross_entropy(out, Y):
...
+Now we have everything we need to start training our neural network.
+We will be training for a target loss of 0.03 with a batch size of 64.
+
+```py
+target_loss = 0.03
+tstep, losst, bs = 0, 1., 64
 
+from tqdm import tqdm
+
+tbar = tqdm(total=0, initial=0, ncols=0, bar_format='Training: {desc}')
```
Comment (wozeparrot) [inline docs/quickstart.md:238]: Probably shouldn't use tqdm to keep this guide from depending on more dependencies then just tinygrad.

File: docs/quickstart.md [hunk 4]
```diff
@@ -135,46 +173,76 @@ def cross_entropy(out, Y):
...
+def run_net(X,Y,bs):
   # random sample a batch
-  samp = np.random.randint(0, X_train.shape[0], size=(64))
-  batch = Tensor(X_train[samp], requires_grad=False)
+  samp = np.random.randint(0, X.shape[0], size=(bs))
+  batch = Tensor(X[samp], requires_grad=False)
   # get the corresponding labels
-  labels = Y_train[samp]
-
+  labels = Y[samp]
```
Comment (wozeparrot) [inline docs/quickstart.md:222]: this shouldn't be in the run_net function as they are not directly related to running the network, they are part of the data loading process.

File: docs/quickstart.md [hunk 5]
```diff
@@ -135,46 +173,76 @@ def cross_entropy(out, Y):
...
+```py
+from tinygrad.nn.optim import get_parameters
+opt = SGD(get_parameters(net), lr=3e-4)
+
+```
 
 Now that we have our network, loss function, and optimizer defined all we are missing is the data to train on!
 There are a couple of dataset loaders in tinygrad located in [/datasets](/datasets).
-We will be using the MNIST dataset loader.
+We will be using the MNIST dataset loader and assign it.
```
Comment (wozeparrot) [inline docs/quickstart.md:205]: What does "assign it" mean in this context?

File: docs/quickstart.md (hunk 2 above)
Comment (foreign-sub) [reply docs/quickstart.md:188]: I did not want to remove it but at the same time it explained what was sent to SGD in the call below so it ended up with a comment

File: docs/quickstart.md (hunk 5 above)
Comment (foreign-sub) [reply docs/quickstart.md:205]: just assign the data to the training and test vars

File: docs/quickstart.md (hunk 2 above)
Comment (wozeparrot) [reply docs/quickstart.md:188]: I think keeping the SGD here and removing the one below would be better because this one is more explicit.

File: docs/quickstart.md (hunk 5 above)
Comment (wozeparrot) [reply docs/quickstart.md:205]: Maybe be more explicit about what variables its assigning to, like `fetch_mnist` returns a tuple with 4 elements, would be better to specify exactly what is being returned and assigned.

File: docs/quickstart.md (hunk 4 above)
Comment (foreign-sub) [reply docs/quickstart.md:222]: i added them to show how much code was shared between running and training in this case otherwise my run_net function was kinda empty and pointless

File: docs/quickstart.md (hunk 3 above)
Comment (foreign-sub) [reply docs/quickstart.md:238]: i used tqdm because it is already a requirement from state.py load_state_dict

File: docs/quickstart.md [hunk 6]
```diff
@@ -135,46 +173,76 @@ def cross_entropy(out, Y):
   y = y.reshape(list(Y.shape)+[num_classes])
   y = Tensor(y)
   return out.mul(y).mean()
+
```
Comment (wozeparrot) [inline docs/quickstart.md:176]: Depending on when/if #1018 is merged this will have to change to match.

File: docs/quickstart.md (hunk 3 above)
Comment (wozeparrot) [reply docs/quickstart.md:238]: Ah OK yea this is fine then.

File: docs/quickstart.md (hunk 5 above)
Comment (foreign-sub) [reply docs/quickstart.md:205]: understood

File: docs/quickstart.md (hunk 1 above)
Comment (foreign-sub) [reply docs/quickstart.md:130]: yes, i'll rephrase it

File: docs/quickstart.md (hunk 2 above)
Comment (foreign-sub) [reply docs/quickstart.md:188]: done

File: docs/quickstart.md (hunk 6 above)
Comment (foreign-sub) [reply docs/quickstart.md:175]: yes

File: docs/quickstart.md [hunk 7]
```diff
@@ -4,11 +4,14 @@ This guide assumes no prior knowledge of pytorch or any other deep learning fram
 It is intended to be a very quick overview of the high level API that tinygrad provides.
 
 This guide is also structured as a tutorial which at the end of it you will have a working model that can classify handwritten digits.
+The code for this tutorial is also available in [quickstart.py](/examples/quickstart.py).
```
Comment (geohot) [inline docs/quickstart.md:7]: Why is it in two places? They should be merged.

Comment (foreign-sub) [reply docs/quickstart.md:7]: merging

File: docs/quickstart.py [hunk 8]
```diff
@@ -79,40 +177,65 @@ def cross_entropy(out, Y):
...
 opt = SGD([net.l1.weight, net.l2.weight], lr=3e-4)
 
+# We can see that we are passing in the parameters of our neural network to the optimizer.
+# This is due to the fact that the optimizer needs to know which parameters to update.
+# The parameters are just listed out explicitly here for clarity.
+# There is a simpler way to do this just by using `get_parameters(net)` from `tinygrad.nn.optim`
+# which will return a list of all the parameters in the neural network like below.
+
 #from tinygrad.nn.optim import get_parameters
 #opt = SGD(get_parameters(net), lr=3e-4)
```
Comment (wozeparrot) [inline docs/quickstart.py]: You can keep this uncommented

File: docs/quickstart.py [hunk 9]
```diff
@@ -79,40 +177,65 @@ def cross_entropy(out, Y):
...
-tbar = tqdm(total=0, initial=0, ncols=0, bar_format='Training: {desc}')
-st = time.perf_counter()
-
-while losst > target_loss:
-  labels, out = run_net(X_train, Y_train, bs)
   # compute loss
-  #loss = cross_entropy(out, labels)
-  loss = sparse_categorical_crossentropy(out, labels)
+  loss = cross_entropy(out, labels)
+  #loss = sparse_categorical_crossentropy(out, labels)
```
Comment (wozeparrot) [inline docs/quickstart.py:238]: I would say just remove this line and add a comment above saying that `cross_entropy` is the same as `sparse_categorical_crossentropy` from `extra.training`

File: docs/quickstart.py (hunk 8 above)
Comment (foreign-sub) [reply docs/quickstart.py]: ok

File: docs/quickstart.py [hunk 10]
```diff
@@ -126,11 +249,11 @@ def run_net(X,Y,bs):
...
   pred = np.argmax(out.numpy(), axis=-1)
   acc = (pred == labels).mean()
-  losst = loss.numpy()
+  step_loss = loss.numpy()
 
   tstep += 1
   if tstep % 100 == 0 or tstep == 1:
-    tbar.desc=(f"Loss: {losst:5.4f} | Accuracy: {acc:5.4f} | Step: {tstep}")
+    tbar.desc=(f"Loss: {step_loss:5.4f} | Accuracy: {acc:5.4f} | Step: {tstep}")
     tbar.update()
```
Comment (wozeparrot) [inline docs/quickstart.py]: can we not use `set_description` here?

**Comment by wozeparrot** [comment]
I'm not actually sure I like this...

This doesn't read as nicely as a formatted markdown doc would, this seems kinda cluttered with broken links as stuff, harder for new people to navigate.

Maybe we just ditch the code file for the quickstart but having it tested in CI to make sure it doesn't break would be nice. Thoughts @geohot?

File: docs/quickstart.py (hunk 9 above)
Comment (foreign-sub) [reply docs/quickstart.py:238]: added comment

File: docs/quickstart.py (hunk 10 above)
Comment (foreign-sub) [reply docs/quickstart.py]: yes we can

**Comment by geohot** [comment]
Honestly, I'm pretty indifferent to this whole thing. It's easy to get carried away in reviews on a PR that just shouldn't happen.

When I see PRs like this that are +500 lines, my instinct is not to merge. Code costs money to maintain, probably on the order of $100 in core tinygrad and $10 elsewhere.

**Comment by foreign-sub** [comment]
yeah it looks big because i have not removed quickstart.md yet, but without it the balance would be more like +20 lines

File: docs/quickstart.md (hunk 4 above)
Comment (foreign-sub) [reply docs/quickstart.md:219]: removed the run_net function

**Comment by foreign-sub** [comment]
reworked it to be only markdown updates and markdown code test passing in ci

**Comment by geohot** [comment]
Ahh, now this has value. What's the minimal change to have the guide pass CI?

**Comment by foreign-sub** [comment]
the markdown fixes

**Comment by foreign-sub** [comment]
repushed with minimal changes

**Comment by geohot** [comment]
Sweet, merged!
